File: Solutions/Problem02.py
# Problem 2: Even Fibonacci Numbers
# find the sum of the even fibonacci numbers less than N
# if N is omitted, 4 million is used
# usage:
# $ python Problem02.py [N = 4000000]
from PEUtils import Fibonacci
import sys

def main():
    N = int(sys.argv[1]) if len(sys.argv) > 1 else 4000000

    # use the Fibonacci generator in PEUtils
    f = Fibonacci()
    v = f.next()
    sum_val = 0

    while v < N:
        if v % 2 == 0:
            sum_val += v
        v = f.next()

    print sum_val

main()

File: RustImplementations/Problem02.rs
/*
  Problem 2: Even Fibonacci Numbers
  Find the sum of the even fibonacci numbers less than 4 million 
*/
extern crate PEUtils;
use PEUtils::Fibonacci;

fn main() {
    let limit = 4000000;   

    // Fibonacci returns a never ending fibonacci sequence 
    let sum = Fibonacci::new()
        .filter(|x| x % 2 == 0)      // filter even
        .take_while(|x| x < &limit)  // below limit
        .fold(0, |b, a| b + a);      // sum

    println!("{}", sum);
}
